### PR TITLE
Chrisdias/linting

### PR DESCRIPTION
**Pull request by chrisdias**
adds linting for dockerfiles using dockerfile_lint library

### CODE REVIEW

File: linting/dockerLinting.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    return;
+}
+
+function doValidate(document: vscode.TextDocument) {
+
+    let diagnostics: vscode.Diagnostic[] = [];
+
+    let linterRuleFile = configOptions.get('linterRuleFile', 'basic_rules.yaml');
+
```
Comment (lostintangent) [inline linting/dockerLinting.ts]: It looks like this lib will actually default to using this exact rules file if you specific a null/empty rules file path, so we could choose not to ship any of these sample rules file, and allow either turning on the default rules or specifying a custom one.

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,3 @@
-out
+out/**/*.js
+out/**/*.map
```
Comment (chrisdias) [inline .gitignore]: actually it does use the .yaml files, they are the linting rules.  i had to change gitignore to persist them. happy if you have another way to do this.

File: linting/dockerLinting.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    if (result.warn.count > 0) {
+        for (let i = 0; i < result.warn.count; i++) {
+            diagnostics.push(new vscode.Diagnostic(new vscode.Range(0, 0, 0, 0), result.warn.data[i].message + ': ' + result.warn.data[i].description, vscode.DiagnosticSeverity.Warning));
+        }
+    }
+
+    diagnosticCollection.set(document.uri, diagnostics);
```
Comment (chrisdias) [inline linting/dockerLinting.ts]: also forgot to do this :)

File: linting/rules/basic_rules_atomic.yaml [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        - 
+          label: "is_latest_tag"
+          regex: /latest/
+          level: "error"
+          message: "base image uses 'latest' tag"
+          description: "using the 'latest' tag may cause unpredictable builds. It is recommended that a specific tag is used in the FROM line or *-released which is the latest supported release."
+          reference_url: 
+            - "https://docs.docker.com/reference/builder/"
+            - "#from"
+          label: "no_tag"
```
Comment (chrisdias) [inline linting/rules/basic_rules_atomic.yaml]: these came with dockerfile_lint so i don't know.  but, we can remove all these per the above comments.

**Comment by chrisdias** [comment]
great feedback!  i've still got work to do, will make changes and then resubmit.

File: dockerExtension.ts [hunk 5]
```diff
@@ -54,4 +58,16 @@ export function activate(ctx: vscode.ExtensionContext): void {
     ctx.subscriptions.push(vscode.commands.registerCommand('vscode-docker.container.open-shell', openShellContainer));
     ctx.subscriptions.push(vscode.commands.registerCommand('vscode-docker.compose.up', composeUp));
     ctx.subscriptions.push(vscode.commands.registerCommand('vscode-docker.compose.down', composeDown));
+
+    diagnosticCollection = vscode.languages.createDiagnosticCollection('docker-diagnostics');
```
Comment (chrisdias) [inline dockerExtension.ts]: fixed

File: linting/rules/basic_rules_atomic.yaml (hunk 4 above)
Comment (chrisdias) [inline linting/rules/basic_rules_atomic.yaml]: removed

File: linting/dockerLinting.ts (hunk 3 above)
Comment (chrisdias) [inline linting/dockerLinting.ts]: fixed

File: linting/dockerLinting.ts (hunk 1 above)
Comment (chrisdias) [inline linting/dockerLinting.ts]: "fixed" by not relying on file...

File: .gitignore (hunk 2 above)
Comment (chrisdias) [inline .gitignore]: fixed

**Comment by lostintangent** [comment]
LGTM!

**Pull request merged**